ユーザー管理機能

miyazawa220618 opened the pull request:
# What
ユーザー管理機能の設定

# What
- ユーザーの新規登録
- 登録済みのユーザーのログイン
- 新規登録・ログインにおけるエラー対処
- ログイン済みのユーザーのログアウト
- 動作確認のための単体テストコード
- Rubocopによる自動修正

### 動作確認キャプチャ
- ユーザーの新規登録ができる動画
https://gyazo.com/990434d8db454b31647cd9063b60acb8
- 入力項目が空のため新規登録でエラーが出る動画
https://gyazo.com/2fc213f3c8d8f81d0b4db46b3b08248a
- 1項目、入力内容に不備があるため新規登録でエラーが出る動画
https://gyazo.com/049e67b34c8718c5fb2377537e1960ba
- ユーザーのログインができる動画
https://gyazo.com/deed01f8dfd3a0ac283bdbd5cca49108
- 入力に不備があるためログインでエラーが出る動画
https://gyazo.com/d72017ea60b26d343ef88afddf06bc04
- ログアウトが成功する動画
https://gyazo.com/9a4325685395f1d9afb3370b145822d4
- テストに成功する画像
https://gyazo.com/4c10eb48237f7789b5d914b12637d948

miyazawa220618 committed:
deviseによるuserテーブルの作成

miyazawa220618 committed:
バリデーションの設定とDLしたviewデータの設置

miyazawa220618 committed:
バリデーションの設定

miyazawa220618 committed:
viewのリンク/formのdummy項目の書き替え

miyazawa220618 committed:
RSpecの設定/exampleの書き出し

miyazawa220618 committed:
userモデルの単体テストコード

miyazawa220618 committed:
RuboCopの自動修正

strongmasahiro commented on app/models/user.rb:
```diff
@@ -0,0 +1,16 @@
+
+  VALID_NAME = /\A[ぁ-んァ-ヶ一-龥々ー]+\z/
+  VALID_KANA = /\A[ァ-ヶー－]+\z/
+
```
以下の実装条件を満たす様にバリデーションを設定しましょう！
```
・パスワードは、半角英数字混合での入力が必須であること。
```

strongmasahiro commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,108 @@
+
+<%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
+<%= form_with model: @user, url: user_registration_path, class: 'registration-main', local: true do |f| %>
+<%# //「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
```
```
<%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
<%# //「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
```
こちらのコメントアウトは、実装案内用のコメントアウトです。
実装完了後、不要になった実装案内用のコメントアウトは全て削除しておきましょう。

理由：
企業様に提出する際に実装案内のコメントが残っていることは好ましくないためです。
※まだ、実装していない部分は残しておいても問題ございません。

strongmasahiro commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,108 @@
+
+  <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
+  <%= render 'shared/error_messages', model: f.object %>
+  <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
```
  <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
  <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
「〜しましょう」というコメントアウトは、実装案内用のコメントアウトです。
実装完了後、不要になった実装案内用のコメントアウトは全て削除しておきましょう。

理由：
企業様に提出する際に実装案内のコメントが残っていることは好ましくないためです。
※まだ、実装していない部分は残しておいても問題ございません。

strongmasahiro commented on db/migrate/20220910092552_add_columns_to_user.rb:
```diff
@@ -0,0 +1,10 @@
+    add_column :users, :first_name, :string
+    add_column :users, :last_name_kana, :string
+    add_column :users, :first_name_kana, :string
+    add_column :users, :birth_date, :date
```
```
    add_column :users, :nickname, :string
    add_column :users, :last_name, :string
    add_column :users, :first_name, :string
    add_column :users, :last_name_kana, :string
    add_column :users, :first_name_kana, :string
    add_column :users, :birth_date, :date
```
追加していただいたカラムに、NOT NULL制約がついておりません。
NOTNULL制約をつけてマイグレーションファイルを修正しましょう！

strongmasahiro commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,17 @@
+
+    nickname              { Faker::Name.name }
+    email                 { Faker::Internet.free_email }
+    password              { Faker::Internet.password(min_length: 6) }
```
```
password              { Faker::Internet.password(min_length: 6) }
```
こちらの記述では、稀に「数字のみ」や「英字のみ」のパスワードが生成されてしまい、テストが失敗することが考えられます。
英数字混合のpasswordを想定した上でFakerを使用する場合は、

```
password              { '1a' + Faker::Internet.password(min_length: 6) }
```
のように、確実にpasswordに英数字が含まれるように指定しましょう！

strongmasahiro commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,104 @@
+      end
+    end
+
+    context '新規登録できない場合' do
```
以下の異常系テストを追加しましょう！
```
・英字のみのパスワードでは登録できない
・数字のみのパスワードでは登録できない
・全角文字を含むパスワードでは登録できない
```

strongmasahiro reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

また、修正後`最新のテスト結果の画像`の添付をお願いいたします。

miyazawa220618 committed:
ユーザー管理機能［修正］

・補助コメントアウトの削除
・追加カラムのnotnill追加
・passwordのバリデーション追加
・passwordのテストコード追加
・rubocopの自動修正

miyazawa220618 commented:
# What
ユーザー管理機能の設定［修正］

# Why
- コメントアウトの削除
- 追加カラムのNOTNILL制約の抜け修正
- passwordのバリデーション追加
- passwordのテストコード追加
- rubocopの自動修正

### 動作確認キャプチャ
- テストに成功する画像
https://gyazo.com/40d1385c048f0d13b7f553d2be2b6948

（バリデーションを追加したため、念の為以下も再度取り直しています。）
- ユーザーの新規登録ができる動画
https://gyazo.com/b2c2d0661bfce74669e239519961007e
- 入力項目が空のため新規登録でエラーが出る動画
https://gyazo.com/d49b57f5376849ccee24a8aeb125fb79

strongmasahiro reviewed (commented):
レビュー依頼ありがとうございます！
LGTMとさせていただきます！実装お疲れ様でした！
引き続き実装頑張っていきましょう！
